File: day3/python/main.py
import re 


def read_data():
    with open('../input.txt') as f:
        data = f.read().replace('\n', '')
    
    return [data]



def part1(data):
    
    pattern = r'mul\((\d{1,3}),\s*(\d{1,3})\)'
    matches = []
    for line in data:
        matches.extend(re.findall(pattern, line))


    sum = 0
    for match in matches:
        sum += int(match[0]) * int(match[1])
    
    return sum




def part2(data):

    do_sections = []
    for line in data:
        parts = line.split('do()')
        # Reached this point, each part looks like this 
        # blah-blah-don't()-blah-blah-blah-don't()-[...]
        for part in parts:
            # We need to get the first part i.e content up to the first don't
            # So we split the part by don't and take the first part
            first_part = part.split('don\'t()')[0]
            do_sections.append(first_part)

    sum = part1(do_sections)
    return sum



if __name__ == '__main__':
    data = read_data()

    #
    print("PART 1")
    print("SOLUTION: ", part1(data))
    #
    print("PART 2")
    print("SOLUTION: ", part2(data))

File: day3/cplusplus/main.cpp
#include <iostream>
#include <fstream>
#include <sstream>
using namespace std;

string read_data(string filename = "../../input.txt")
{
    ifstream file(filename);
    stringstream buffer;
    buffer << file.rdbuf();
    string s = buffer.str();

    // add padding
    return "################" + s + "################";
}

int getDigits(const string &data, int &i)
{
    string num = "";
    while (i < data.length() && isdigit(data[i]))
    {
        num += data[i];
        i++;
    }
    return num.empty() ? -1 : stoi(num);
}

int part1(string data)
{
    int n = (int)data.length();
    int sum = 0;
    for (int i = 0; i < n; i++)
    {
        if (data.substr(i, 4) == "mul(")
        {
            i += 4;
            int x = getDigits(data, i);
            if (data[i] == ',')
            {
                i++;
                int y = getDigits(data, i);
                if (data[i] == ')')
                {
                    if (x != -1 && y != -1)
                    {
                        sum += x * y;
                    }
                }
            }
        }
    }
    return sum;
}

int part2(string data)
{
    int n = (int)data.length();
    int sum = 0;
    bool enabled = true;
    for (int i = 0; i < n; i++)
    {
        // do check
        if (data.substr(i, 4) == "do()" && !enabled)
        {
            enabled = true;
            i += 3;
            continue;
        }

        // don't check
        if (data.substr(i, 7) == "don't()" && enabled)
        {
            enabled = false;
            i += 6;
            continue;
        }

        if (enabled && data.substr(i, 4) == "mul(")
        {
            i += 4;
            int x = getDigits(data, i);
            if (data[i] == ',')
            {
                i++;
                int y = getDigits(data, i);
                if (data[i] == ')')
                {
                    if (x != -1 && y != -1)
                    {
                        sum += x * y;
                    }
                }
            }
        }
    }
    return sum;
}

int main()
{
    string data = read_data();

    cout << "Part 1: " << part1(data) << endl;
    cout << "Part 2: " << part2(data) << endl;

    return 0;
}